src/_nebari/stages/kubernetes_services/template/modules/kubernetes/services/jupyterhub/files/jupyterhub/04-auth.py

```diff
@@ -154,17 +164,36 @@ async def load_managed_roles(self):
 
         return list(roles.values())
 
-    def _get_scope_from_role(self, role):
+    def _get_scope_from_role(self, role, component_filter=["jupyterhub"]):
         """Return scopes from role if the component is jupyterhub"""
         role_scopes = role.get("attributes", {}).get("scopes", [])
         component = role.get("attributes", {}).get("component")
         # Attributes are returned as a single-element array, unless `##` delimiter is used in Keycloak
         # See this: https://stackoverflow.com/questions/68954733/keycloak-client-role-attribute-array
-        if component == ["jupyterhub"] and role_scopes:
+        if component == component_filter and role_scopes:
             return self.validate_scopes(role_scopes[0].split(","))
+        elif component_filter is None:
+            return self.validate_scopes(role_scopes)
         else:
             return []
 
+    async def _fetch_and_map_roles(self, roles, token, url, group_name_key="path"):
+        # we could use either `name` (e.g. "developer") or `path` ("/developer");
+        # since the default claim key returns `path`, it seems preferable.
+
+        self.log.info("Mapping roles with groups and users..")
+
+        for role in roles:
+            role_name = role["name"]
+            # fetch role assignments to groups
+            base_url = url.format(role_name=role_name)
+            groups = await self._fetch_api(f"{base_url}/groups", token=token)
+            role["groups"] = [group[group_name_key] for group in groups]
+            users = await self._fetch_api(f"{base_url}/users", token=token)
+            role["users"] = [user["username"] for user in users]
+
+        return roles
```

Comment:
The `groups` and `users` are already available in `user_roles_rich`, which means this shouldn't be required.